src/main/java/com/microsoft/aad/msal4j/TokenCache.java

```diff
@@ -300,16 +302,92 @@ static private AppMetadataCacheEntity createAppMetadataCacheEntity(TokenRequestE
+                    IdTokenCacheEntity homeAcctIdToken = idTokens.get(getIdTokenKey(
+                            homeAccCached.homeAccountId(),
+                            homeAccCached.environment(),
+                            clientId,
+                            homeAccCached.realm()));
+
+                    Account homeAcc = (Account) homeAccCached.toAccount();
+                    homeAcc.idTokenClaims = JWTParser.parse(homeAcctIdToken.secret).getJWTClaimsSet().getClaims();
+
+                    Iterator<AccountCacheEntity> tenantAcctsIterator = tenantAccounts.iterator();
+                    Map<String, ITenantProfile> tenantProfiles = new HashMap<>();
+
+                    while (tenantAcctsIterator.hasNext()) {
+                        AccountCacheEntity tenantAccCached = tenantAcctsIterator.next();
+                        //If the tenant account's home ID matches a home ID (UID) in the list of home accounts, it is a
+                        //tenant profile of that home account
+                        if (tenantAccCached.homeAccountId.equals(homeAccCached.homeAccountId())) {
+                            IdTokenCacheEntity token = idTokens.get(getIdTokenKey(
+                                    tenantAccCached.homeAccountId(),
+                                    tenantAccCached.environment(),
+                                    clientId,
+                                    tenantAccCached.realm()));
+
+                            ITenantProfile profile = new TenantProfile(tenantAccCached.localAccountId(),
+                                            tenantAccCached.realm(),
+                                            JWTParser.parse(token.secret()).getJWTClaimsSet().getClaims());
+                            tenantProfiles.put(tenantAccCached.realm(), profile);
+                            tenantAcctsIterator.remove();
+                        }
+                    }
+
+                    if (tenantProfiles.size() > 0) {
+                        homeAcc.tenantProfiles(tenantProfiles);
+                    }
+
+                    rootAccounts.add(homeAcc);
+                }
+                //Create Accounts from any tenant profile which did not have a corresponding home account,
+                // and add it to the list of root accounts
+                for (AccountCacheEntity tenantProfileNoHome : tenantAccounts) {
+                    IdTokenCacheEntity token = idTokens.get(getIdTokenKey(
+                            tenantProfileNoHome.homeAccountId(),
+                            tenantProfileNoHome.environment(),
+                            clientId,
+                            tenantProfileNoHome.realm()));
+                    Account account = (Account) tenantProfileNoHome.toAccount();
+                    account.idTokenClaims(JWTParser.parse(token.secret).getJWTClaimsSet().getClaims());
+                    rootAccounts.add(account);
+                }
+
+                return rootAccounts;
```

Comment:
the logic of aggregation seem a bit complicated, why would not you just fill 
map <homeAccountId, Set<AccountCacheEntity> > than iterate it and populate tenantProfiles ?
that would be easier to read + complexity will be O(N) instead of O(N pow 2)

Reply:
We need the 'realm' value of an AccountCacheEntity to know which ID tokens are relevant to an Account, and this value is not added to the Account class. Would it be alright to just add a new `realm` field to the Account class? If so, then I could turn all of the home AccountCacheEntities to Accounts, and just iterate over the tenant profiles like you suggest.